README
The best way to find out about these toys is to try them out; they're
all small demos that have few/no dependencies beyond the clutter
libraries. The following is a summary of each;

aaina         - A Flickr picture browser
astro-desktop - A mock-up desktop/palmtop application environment
courasel      - A carousel-style menu demo
foofone       - A mock-up mobile phone interface
mallums-magic-browser - A webkit-clutter based web browser
opt           - A slide-show application
ripples       - A small 'ripples' effect demo (using the cogl
vector-drawing API)
sqlite-model  - an sqlite-backed ClutterModel implementation
widgets       - A texture-reflecting actor demo
youhaa        - A (currently broken) YouTube video browser
circles       - A small spinning circles demo (using the cogl
vector-drawing API)
fluttr        - Another Flickr picture browser
gcr           - A Gegl-based clutter video recorder
odo           - A mesh deformation demo set
pong          - An unfinished pong game
script-viewer - A ClutterScript viewer
table         - A multi-touch table style picture/video viewer (not
actually multi-touch)
woohaa        - A video browser/player



opt/ChangeLog
2008-10-13  Matthew Allum  <mallum@openedhand.com>

	* configure.ac:
        Use clutter 0.8
	* opt-show.c:
	* opt-slide.c:
        Dont hardcode number of channels when setting texture data,
        use gdk_pixbuf_get_n_channels ()
	* opt-transition.c: (fade_transition_frame_cb):
        Fix fade. Timelines start from frame #1 now.

2008-07-01  Emmanuele Bassi  <ebassi@openedhand.com>

	* configure.ac:
	* opt-config.c:
	* opt-show.c:
	* opt-slide.c:
	* opt.c: Update to clutter-0.7 to test out the code base.

2008-02-21  Øyvind Kolås  <pippin@o-hand.com>

	* opt-config.c: (opt_config_load): permuted arguments of memset
	to make it correct.

2008-02-18  Chris Lord  <chris@openedhand.com>

	* configure.ac:
	Bump clutter version to 0.6

2008-02-08  Chris Lord  <chris@openedhand.com>

	* configure.ac:
	* opt-menu.c: (opt_menu_new):
	* opt-show.c: (transition_completed_cb):
	* opt-transition.c: (yz_flip_transition_frame_cb),
	(zoom_transition_frame_cb), (flip_transition_frame_cb),
	(cube_transition_frame_cb), (page_transition_frame_cb):
	Update to use 0.5

2007-08-07  Matthew Allum  <mallum@openedhand.com>

	* configure.ac:
	Update to use 0.4

2007-06-19  Matthew Allum  <mallum@openedhand.com>

	* demo.xml:
	Include cube transition

	* opt-show.c:
	Add png compression parameter and set at maximum

	* opt.c:
	Call clutter_init() with args

2007-06-08  Matthew Allum  <mallum@openedhand.com>

	* configure.ac:
	* opt-menu.c:
	* opt-show.c:
	* opt-transition.c:
	* opt.c: 
	Update to 0.3 API, thanks to Johan Bilien  (#366).
	Print error if offscreen stage unsupported.

2007-04-16  Tomas Frydrych  <tf@openedhand.com>

	* opt-menu.c:
	Changed scaling gravity from SW to NW.
	
2007-03-29  Matthew Allum  <mallum@openedhand.com>

	* Makefile:
	Remove.
	* Makefile.am:
	* autogen.sh:
	* configure.ac:
	Autofooify.

	* demo.xml:
	Add a simple demo show with instructions.

	* opt-show.c:
	Improve export to generate simple XML.

2007-02-28  Tomas Frydrych  <tf@openedhand.com>

	* Makefile:
	Put clutter version requirement back to 0.2
	
	* opt-menu.c:
	Animated menu
	
2007-02-28  Tomas Frydrych  <tf@openedhand.com>

	* opt-transition.h:
	* opt-transition.c:
	* opt-config.c:
	Added new transition type 'page'

	* Makefile:
	* opt.h:
	* opt-menu.h:
	* opt-menu.c:
	* opt-show.h:
	* opt-show.c:
	Added simple menu

	* opt.c: (input_cb): 
	Added 'm' to bring up menu, explicitely Right to advance, button 3
	to retreat; ignore Up, Down, Return and buttons 2, 4, 5 used by
	menu.
	
2007-01-26  Matthew Allum  <mallum@openedhand.com>

	* opt.doap:
	Add a doap file

2007-01-17  Matthew Allum  <mallum@openedhand.com>

	* Makefile:
	* opt-show.c:
	* opt-slide.c: (opt_slide_set_title), (opt_slide_add_bullet_text_item):
	Update for 0.2 API changes.

2006-07-03  Ross Burton  <ross@openedhand.com>

	* opt-slide.c:
	Remove unused constants.

2006-07-03  Ross Burton  <ross@openedhand.com>

	* opt-show.c:
	Use somewhat better filenames when exporting. Should really count
	slides to use the right number but this works for most
	presentations.

2006-07-03  Ross Burton  <ross@openedhand.com>

	* opt.dtd:
	Allow slides with only titles (Wouter Bolsterlee <uws@xs4all.nl>)

2006-06-23  Matthew Allum  <mallum@openedhand.com>

	* opt-transition.c: (zoom_transition_frame_cb):
	Fix zoom transition as to not leave next frame 180 degrees

2006-06-22  Ross Burton  <ross@openedhand.com>

	* Makefile:
	Use clutter-0.1.

	* opt.dtd:
	Add zoom transition.

2006-06-21  Emmanuele Bassi  <ebassi@openedhand.com>

	* opt-show.h:
	* opt-show.c (opt_show_update_position_label),
	(transition_completed_cb), (opt_show_toggle_position): Toggle
	an overlayed rectangle containing the current slide and the
	total number of slides.

	* opt.c (input_cb): Add binding for showing the slide number.

2006-06-20  Ross Burton  <ross@openedhand.com>

	* opt-transition.c:
	Don't hard-code frame count anywhere but the creation, get it from
	the timeline.

2006-06-20  Ross Burton  <ross@openedhand.com>

	* opt-transition.c:
	Add a define for the transition midpoint frame, and hook up the
	zoom transition.

2006-06-20  Ross Burton  <ross@openedhand.com>

	* opt-config.c:
	Scale background images to the stage size when loading to avoid
	scaling at runtime.

2006-06-19  Matthew Allum  <mallum@openedhand.com>

	* opt-config.c:
	* opt-show.c:
	* opt-slide.c:
	* opt-slide.h:
	Integrate Mr Burtons per slide background patch.
	Pre-realize textures before a transition.
	Correctly bail on error if image cannot be loaded.

2006-06-19  Matthew Allum  <mallum@openedhand.com>

	* opt-config.c:
	* opt-transition.c: (yz_flip_transition_frame_cb):
	* opt-transition.h:
	* test.xml:
	Add new zoom transition. 

2006-06-19  Emmanuele Bassi  <ebassi@openedhand.com>

	* opt-show.c:
	Add support for the PgUp and PgDown keys, for skipping 5 slides
	backward and forward the current position.

2006-06-19  Ross Burton  <ross@openedhand.com>

	* opt-show.c:
	Marshal the background property as a GdkPixbuf object not a pointer.

2006-06-19  Ross Burton  <ross@openedhand.com>

	* hirez/oh-present.xcf:
	Title the layers and remove empty/duplicate layers.

2006-06-19  Ross Burton  <ross@openedhand.com>

	* opt-transition.c:
	Change the rotation of the cube when switching.

2006-06-19  Ross Burton  <ross@openedhand.com>

	* opt.dtd:
	Image tags have no content.

2006-06-15  Matthew Allum  <mallum@openedhand.com>

	* opt.c: (input_cb):
	Keep Tomas happy with support with mouse button 
	slide forwards/back.

2006-06-14  Matthew Allum  <mallum@openedhand.com>

	* opt-config.c:
	* opt-show.c:
	* opt-show.h:
	Fix bullet symbol color.

2006-06-14  Emmanuele Bassi  <ebassi@openedhand.com>

	* opt.c (input_cb): Use clutter_main_quit() instead of
	calling exit().

2006-06-14  Ross Burton  <ross@openedhand.com>

	* opt.dtd:
	Remove img from <defaults> and set valid enumeration for symbol
	and style.

2006-06-13  Matthew Allum  <mallum@openedhand.com>

	* opt-config.c:
	* opt.dtd:
	* test.xml:

odo/odo.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <glib.h>

#include <clutter/clutter.h>
#include "clutter-texture-odo.h"

static int tile_width  = 8;
static int tile_height = 8;
static char * image = "redhand.png";

struct distort_data
{
    ClutterFixed t;
    ClutterActor * tex;
};

void new_frame_cb (ClutterTimeline *timeline,
                   gint             frame_num,
                   gpointer         data)
{
    struct distort_data * d = data;
    guint frames = clutter_timeline_get_n_frames (timeline);
    gdouble t = 0.7 - ((gdouble)abs(frames/2-frame_num))*1.4 / (gdouble)frames;

    d->t = CLUTTER_FLOAT_TO_FIXED (t);

    clutter_actor_queue_redraw (clutter_stage_get_default ());
}

/* skew paramter t <0, 0.7>, the greater t, the greater distortion */
static gboolean
distort_func1 (ClutterTextureOdo * texture,
               ClutterFixed x, ClutterFixed y, ClutterFixed z,
               ClutterFixed *x2, ClutterFixed *y2, ClutterFixed *z2,
               ClutterColor *color, gpointer data)
{
    struct distort_data * d = data;
    gint w, h;

    clutter_texture_get_base_size (CLUTTER_TEXTURE (d->tex), &w, &h);

    *x2 = x;
    *y2 = y;
    *z2 =  z + CFX_MUL (d->t,
                        w * clutter_sini (CLUTTER_FIXED_TO_INT (x)*1024/w));
    
    return TRUE;
}

/* skew paramter t <0, 0.7>, the greater t, the greater distortion */
static gboolean
distort_func2 (ClutterTextureOdo * texture,
               ClutterFixed x, ClutterFixed y, ClutterFixed z,
               ClutterFixed *x2, ClutterFixed *y2, ClutterFixed *z2,
               ClutterColor *color, gpointer data)
{
  struct distort_data * d = data;
  gint w, h, xi, yi, zi;

  clutter_texture_get_base_size (CLUTTER_TEXTURE (d->tex), &w, &h);

  xi = CLUTTER_FIXED_TO_INT (x);
  yi = CLUTTER_FIXED_TO_INT (y);
  zi = CLUTTER_FIXED_TO_INT (z);
  
  *x2 = x + CFX_MUL(d->t * w/2, clutter_cosi(xi*512/w)) - (d->t * w/2);
    
  *y2 = y + CFX_MUL(d->t * h/2, clutter_cosi(yi*512/h)) - (d->t * h/2);

  *z2 = z - CFX_MUL (d->t, w * clutter_sini (xi*512/w));
  
  return TRUE;
}

/* skew paramter t <0, 0.7>, the greater t, the greater distortion */
static gboolean
distort_func3 (ClutterTextureOdo * texture,
               ClutterFixed x, ClutterFixed y, ClutterFixed z,
               ClutterFixed *x2, ClutterFixed *y2, ClutterFixed *z2,
               ClutterColor *color, gpointer data)
{
  struct distort_data * d = data;
  gint w, h, xi, yi;

  clutter_texture_get_base_size (CLUTTER_TEXTURE (d->tex), &w, &h);

  xi = CLUTTER_FIXED_TO_INT (x);
  yi = CLUTTER_FIXED_TO_INT (y);

  *x2 = x + CFX_MUL(d->t * w/2, clutter_cosi(xi*512/w)) - (d->t * w/2);
    
  *y2 = y + CFX_MUL(d->t * h/2, clutter_cosi(yi*512/h)) - (d->t * h/2);

  *z2 =  z;
  
  return TRUE;
}

static void 
on_event (ClutterStage *stage,
          ClutterEvent *event,
          gpointer      data)
{
  switch (event->type)
    {
    case CLUTTER_KEY_PRESS:
      {
        guint sym = clutter_key_event_symbol ((ClutterKeyEvent*)event);

        switch (sym)
          {
          case CLUTTER_Escape:
          case CLUTTER_q:
          case CLUTTER_Q:
             clutter_main_quit ();
             break;
             
          default:
            break;
          }
      }
      break;
    default:
      break;
    }
}


static ClutterActor * beget_odo (ClutterTexture * tex,
                                 gint x,
                                 gint y,
                                 gdouble rotate_x,
                                 gdouble rotate_y,
                                 ClutterTextureDistortFunc func,
                                 gpointer data)
{
  ClutterActor * odo = clutter_texture_odo_new (tex);
  clutter_actor_set_position (odo, x, y);
  clutter_actor_set_rotation (odo, CLUTTER_X_AXIS, rotate_x, 0, 0, 0);
  clutter_actor_set_rotation (odo, CLUTTER_Y_AXIS, rotate_y, 0, 0, 0);

  g_object_set (G_OBJECT (odo), "tile-width", tile_width, NULL);
  g_object_set (G_OBJECT (odo), "tile-height", tile_height, NULL);
  g_object_set (G_OBJECT (odo), "distort-func-data", data, NULL);
  g_object_set (G_OBJECT (odo), "distort-func", func, NULL);
  return odo;
}

int
main (int argc, char *argv[])
{
  ClutterActor     *stage;
  ClutterActor     *hand, *odo;
  ClutterColor      stage_color = { 0xcc, 0xcc, 0xcc, 0xff };
  struct distort_data data;
  ClutterTimeline  *timeline;
  gint              i;
  ClutterMesh       mesh;
  
  for (i = 1; i < argc; ++i)
    {
      if (!strncmp (argv[i], "--tile-height", 13))
        {
          tile_height = atoi (argv[i] + 14);

          if (!tile_height)
            tile_height = 8;
        }
      else if (!strncmp (argv[i], "--tile-width", 12))
        {
          tile_width = atoi (argv[i] + 13);

          if (!tile_width)
            tile_height = 8;
        }
      else if (!strncmp (argv[i], "--image", 7))
          image = argv[i] + 8;
    }
  
  clutter_init (&argc, &argv);

  stage = clutter_stage_get_default ();

  g_signal_connect (stage, "key-press-event",
                    G_CALLBACK (on_event),
                    NULL);

  clutter_stage_set_color (CLUTTER_STAGE (stage),
                           &stage_color);

  clutter_actor_set_size (stage, 800, 600);

  /* Make a hand */
  hand = clutter_texture_new_from_file (image, NULL);
  clutter_actor_show (hand);

  data.t = CLUTTER_FLOAT_TO_FIXED (0.7);
  data.tex = hand;
  
  odo = beget_odo (CLUTTER_TEXTURE (hand),
                   20, 20,
                   15.0, 30.0,
                   distort_func1,
                   &data);
                   
  clutter_container_add (CLUTTER_CONTAINER (stage), odo, NULL);

  odo = beget_odo (CLUTTER_TEXTURE (hand),
                   570, 400,
                   340.0, 0.0,
                   distort_func2,
                   &data);
                   
  clutter_container_add (CLUTTER_CONTAINER (stage), odo, NULL);

  odo = beget_odo (CLUTTER_TEXTURE (hand),
                   600, 20,
                   0.0, 0.0,
                   distort_func3,
                   &data);
                   
  clutter_container_add (CLUTTER_CONTAINER (stage), odo, NULL);

  mesh.dimension_x = 4;
  mesh.dimension_y = 4;
  mesh.points = g_new (ClutterMeshPoint, 16);

#define P(a,b,_x,_y,_z) \
   mesh.points[b*mesh.dimension_x + a].x = CLUTTER_INT_TO_FIXED(_x); \
   mesh.points[b*mesh.dimension_x + a].y = CLUTTER_INT_TO_FIXED(_y); \
   mesh.points[b*mesh.dimension_x + a].z = CLUTTER_INT_TO_FIXED(_z)

  P (0,0,  0,0,0);
  P (0,1, 80,0,0);
  P (0,2,160,0,0);
  P (0,3,240,0,0);

  P (1,0,  0,80,0);
  P (1,1, 80,80,160);
  P (1,2,160,80,160);
  P (1,3,240,80,0);

  P (2,0,  0,160,0);
  P (2,1, 80,160,160);
  P (2,2,160,160,160);
  P (2,3,240,160,0);

  P (3,0,  0,240,0);
  P (3,1, 80,240,0);
  P (3,2,160,240,0);
  P (3,3,240,240,0);
#undef P
  
  odo = clutter_texture_odo_new (CLUTTER_TEXTURE (hand));
  clutter_actor_set_position (odo, 290, 200);

  g_object_set (G_OBJECT (odo), "mesh", &mesh, NULL);
  
  clutter_container_add (CLUTTER_CONTAINER (stage), odo, NULL);

  timeline = clutter_timeline_new (106, 26);
  clutter_timeline_set_loop (timeline, TRUE);
  
  g_signal_connect (timeline, "new-frame", G_CALLBACK (new_frame_cb), &data);
  
  clutter_actor_show_all (stage);

  clutter_timeline_start (timeline);
  
  clutter_main();

  return 0;
}
